### PR TITLE
Support n-ary monotonic functions in ordering equivalence

### LINKED ISSUE #13839: Support n-ary monotonic functions in ordering equivalence
### Is your feature request related to a problem or challenge?

Currently, DataFusion's ordering equivalence only works with single-parameter monotonic functions. We need to support functions with multiple parameters to enable more optimization opportunities.
        
Example:
- Table ordered by `[c ASC, a ASC, b ASC]` with equality `c = concat(a, b)`
- Since `concat()` preserves lexicographical order, then `[a ASC, b ASC]` should be valid ordering

### Describe the solution you'd like

Add support for N-ary monotonic functions by:
1. Add `preserves_lex_ordering` property to ExprProperties
2. Implement for `concat()` function
3. Update ordering equivalence logic to handle multiple parameters

### Describe alternatives you've considered

_No response_

### Additional context

This will improve query optimization by allowing more valid orderings to be discovered.

### PR DESCRIPTION
**Pull request by gokselk**
## Which issue does this PR close?

Closes #13839.

## Rationale for this change

DataFusion's ordering equivalence system currently only handles single-parameter monotonic functions. This PR extends support to N-ary functions that preserve lexicographical ordering, like concat(). This enables more optimization opportunities when working with ordered data.

## What changes are included in this PR?

- Added `preserves_lex_ordering` field to `ExprProperties`
- Implemented `output_preserves_lex_ordering` for concat function
- Updated `discover_new_orderings()` to handle multiple parameters
- Added tests for concat and non-monotonic multiply cases

## Are these changes tested?

Yes, added three new test cases:
- test_ordering_equivalence_with_monotonic_concat
- test_ordering_equivalence_with_non_monotonic_multiply
- test_ordering_equivalence_with_concat_equality

## Are there any user-facing changes?

No user-facing changes. This is an internal optimization improvement.

### CODE REVIEW

**Comment by gokselk** [comment]
cc: @berkaysynnada @ozankabak

**Comment by gokselk** [comment]
### Note about Cargo.toml changes

I added `datafusion-functions` as a dependency to `datafusion/physical-expr/Cargo.toml` to use the concat function in tests.

File: datafusion-cli/Cargo.lock [hunk 1]
```diff
@@ -1549,6 +1549,7 @@ dependencies = [
  "datafusion-common",
  "datafusion-expr",
  "datafusion-expr-common",
+ "datafusion-functions",
```
Comment (berkaysynnada) [inline datafusion-cli/Cargo.lock:1552]: I guess this is leftover

File: datafusion/physical-expr/Cargo.toml [hunk 2]
```diff
@@ -44,6 +44,7 @@ arrow-schema = { workspace = true }
 datafusion-common = { workspace = true, default-features = true }
 datafusion-expr = { workspace = true }
 datafusion-expr-common = { workspace = true }
+datafusion-functions = { workspace = true }
```
Comment (jayzhan211) [inline datafusion/physical-expr/Cargo.toml:47]: since only test requires this, maybe move to dev-dependencies

Comment (jayzhan211) [reply datafusion/physical-expr/Cargo.toml:47]: or maybe move the test in `core/test` 🤔

Comment (gokselk) [reply datafusion/physical-expr/Cargo.toml]: I've moved it to dev-dependencies for now

File: datafusion/expr/src/udf.rs [hunk 3]
```diff
@@ -809,6 +837,10 @@ impl ScalarUDFImpl for AliasedScalarUDFImpl {
         self.inner.output_ordering(inputs)
     }
 
+    fn output_preserves_lex_ordering(&self, inputs: &[ExprProperties]) -> Result<bool> {
```
Comment (ozankabak) [inline datafusion/expr/src/udf.rs:840]: Seems like you don't need the `output_` prefix, just `preserves_lex_ordering` is enough.

File: datafusion/expr/src/udf.rs [hunk 4]
```diff
@@ -652,8 +659,29 @@ pub trait ScalarUDFImpl: Debug + Send + Sync {
...
+            .all(|input| &input.sort_properties == first_order)
+        {
+            Ok(*first_order)
+        } else {
+            Ok(SortProperties::Unordered)
+        }
+    }
+
+    /// Whether the function preserves lexicographical ordering based on the input ordering
+    fn output_preserves_lex_ordering(&self, _inputs: &[ExprProperties]) -> Result<bool> {
```
Comment (ozankabak) [inline datafusion/expr/src/udf.rs:683]: Seems like you don't need the `output_` prefix, just `preserves_lex_ordering` is enough.

**Review by ozankabak — APPROVED** [review]
LGTM, just a naming comment

File: datafusion/expr/src/udf.rs (hunk 4 above)
Comment (gokselk) [reply datafusion/expr/src/udf.rs]: Resolved in 91c7012.

File: datafusion/expr/src/udf.rs (hunk 3 above)
Comment (gokselk) [reply datafusion/expr/src/udf.rs]: Resolved in 91c7012.

**Comment by berkaysynnada** [comment]
will be merged once CI passes one more time

**Pull request merged**